components/browser/session/src/main/java/mozilla/components/browser/session/storage/SessionStorage.kt

```diff
@@ -4,59 +4,428 @@
-    fun start(sessionManager: SessionManager)
+    fun whenGoingToBackground(
+        lifecycle: Lifecycle = ProcessLifecycleOwner.get().lifecycle
+    ): AutoSave {
+        lifecycle.addObserver(AutoSaveBackground(this))
+        return this
+    }
 
     /**
-     * Stops persisting the state automatically.
+     * Saves the state automatically when the sessions change, e.g. sessions get added and removed.
      */
-    fun stop()
+    fun whenSessionsChange(): AutoSave {
+        AutoSaveSessionChange(
+            this,
+            sessionManager
+        ).observeSelected()
+        return this
+    }
+
+    @Synchronized
+    internal fun triggerSave(): Job {
+        val currentJob = saveJob
+
+        if (currentJob != null && currentJob.isActive) {
+            logger.debug("Skipping save, other job already in flight")
+            return currentJob
+        }
+
+        val now = now()
+        val delayMs = lastSaveTimestamp + minimumIntervalMs - now
+        lastSaveTimestamp = now
+
+        GlobalScope.launch(Dispatchers.IO) {
+            if (delayMs > 0) {
+                logger.debug("Delaying save (${delayMs}ms)")
+                delay(delayMs)
+            }
+
+            val start = now()
+
+            try {
+                val snapshot = sessionManager.createSnapshot()
+                if (snapshot != null) {
+                    sessionStorage.save(snapshot)
+                }
+            } finally {
+                val took = now() - start
+                logger.debug("Saved state to disk [${took}ms]")
+            }
+        }.also {
+            saveJob = it
+            return it
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun now() = SystemClock.elapsedRealtime()
+}
+
+/**
+ * [LifecycleObserver] to start/stop a task that saves the state at a periodic interval.
+ */
+private class AutoSavePeriodically(
+    private val autoSave: AutoSave,
+    private val scheduler: ScheduledExecutorService,
+    private val interval: Long,
+    private val unit: TimeUnit
+) : LifecycleObserver {
+    private var scheduledFuture: ScheduledFuture<*>? = null
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_START)
+    fun start() {
+        scheduledFuture = scheduler.scheduleAtFixedRate(
+            {
+                autoSave.logger.info("Save: Periodic")
+                autoSave.triggerSave()
+            },
+            interval,
+            interval,
+            unit
+        )
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun stop() {
+        scheduledFuture?.cancel(false)
+    }
+}
+
+/**
+ * [LifecycleObserver] to save the state when the app goes to the background.
+ */
+private class AutoSaveBackground(
+    private val autoSave: AutoSave
+) : LifecycleObserver {
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun stop() {
+        autoSave.logger.info("Save: Background")
+
+        runBlocking {
+            autoSave.triggerSave().join()
```

Comment:
Do we need a flag here or similar to prevent a potential scheduled delay? If a (periodic) save happened less than 2 seconds (minimum interval) ago, and the app goes to the background we would delay the save, right? So, we'd be blocking for longer than necessary.

Reply:
Oh yes, that's a good point. I'll update the PR to add some kind of "no delay flag".